service/frontend/health_check.go

```diff
@@ -91,17 +95,40 @@ func (h *healthCheckerImpl) Check(ctx context.Context) (enumsspb.HealthState, er
 	}
 
 	var failedHostCount float64
+	var hostDeclinedServingCount float64
 	for i := 0; i < len(hosts); i++ {
 		healthState := <-receiveCh
-		if healthState == enumsspb.HEALTH_STATE_NOT_SERVING {
+		switch healthState {
+		case enumsspb.HEALTH_STATE_NOT_SERVING, enumsspb.HEALTH_STATE_UNSPECIFIED:
 			failedHostCount++
+		case enumsspb.HEALTH_STATE_DECLINED_SERVING:
+			hostDeclinedServingCount++
+		case enumsspb.HEALTH_STATE_SERVING:
+			// Do nothing.
 		}
 	}
 	close(receiveCh)
 
-	if (failedHostCount / float64(len(hosts))) > h.hostFailurePercentage() {
+	// Make sure that at lease 2 hosts must be not ready to trigger this check.
+	proportionOfDeclinedServiceHosts := ensureMinimumProportionOfHosts(h.hostDeclinedServingProportion(), len(hosts))
+
+	hostDeclinedServingProportion := hostDeclinedServingCount / float64(len(hosts))
+	if hostDeclinedServingProportion > proportionOfDeclinedServiceHosts {
+		h.logger.Warn("health check exceeded host declined serving proportion threshold", tag.NewFloat64("host declined serving proportion threshold", proportionOfDeclinedServiceHosts))
+		return enumsspb.HEALTH_STATE_DECLINED_SERVING, nil
+	}
+
+	failedHostCountProportion := failedHostCount / float64(len(hosts))
+	if failedHostCountProportion+hostDeclinedServingProportion > h.hostFailurePercentage() {
```

Comment:
I am a little worry about false positive:
1. double count in failure 
2. the minimum proportion

Reply:
The double count in failure is a good point. If we make changes to the way we mark pods as unhealthy, then this could overlap. Currently, I'm not sure that we could both have persistence latency/errors, and also be in this not serving state, because this internal hostDeclinedServing value is only set on startup of the host (we don't currently flip this back to unhealthy), so no persistence errors exist. 

To the second point, I think that a host not serving traffic at all should factor into our hostFailurePercentage to catch the issue from before where we had some errors, but most pods were crashlooping so we didn't catch this case. If some hosts are failing, it means we have to pick up the slack on other hosts, so it's "failing". WDYT?